update-engine/src/engine.rs

```diff
@@ -0,0 +1,779 @@
+#[derive(Debug)]
+pub struct NewStep<'engine, 'a, S: StepSpec, T> {
+    steps: &'engine Mutex<Steps<'a, S>>,
+    component: S::Component,
+    id: S::StepId,
+    description: Cow<'static, str>,
+    exec_fn: DebugIgnore<StepExecFn<'a, S>>,
+    receiver: oneshot::Receiver<T>,
+    metadata_fn: Option<DebugIgnore<StepMetadataFn<'a, S::StepMetadata>>>,
+}
+
+impl<'engine, 'a, S: StepSpec, T> NewStep<'engine, 'a, S, T> {
+    /// Adds a metadata-generating function to the step.
+    ///
+    /// This function is expected to produce [`S::StepMetadata`]. The metadata
+    /// function must be infallible, and will often just be synchronous code.
+    pub fn with_metadata_fn<F, Fut>(mut self, f: F) -> Self
+    where
+        F: FnOnce() -> Fut + 'a,
+        Fut: Future<Output = S::StepMetadata> + Send + 'a,
+    {
+        self.metadata_fn = Some(DebugIgnore(Box::new(|| (f)().boxed())));
+        self
+    }
+
+    /// Registers the step with the engine.
+    pub fn register(self) -> StepHandle<T> {
+        let mut steps_lock = self.steps.lock().unwrap();
+        let component_count = steps_lock
+            .component_counts
+            .entry(self.component.clone())
+            .or_insert(0);
+        let current_index = *component_count;
+        *component_count += 1;
+
+        let step = Step {
+            metadata_gen: StepMetadataGen {
+                id: self.id,
+                component: self.component.clone(),
+                component_index: current_index,
+                description: self.description,
+                metadata_fn: self.metadata_fn,
+            },
+            exec: StepExec { exec_fn: self.exec_fn },
+        };
+        steps_lock.steps.push(step);
+        StepHandle { receiver: self.receiver }
+    }
+}
+
+pin_project! {
+    /// A future that resolves to a step's result.
+    ///
+    /// This handle can be used between steps to transfer data between steps.
+    ///
+    /// A `StepHandle` implements `Future`, so it can be awaited. To share the
+    /// result across several dependent futures, use
+    /// [`FutureExt::shared`](futures::FutureExt::shared).
+    ///
+    /// It is important that `StepHandle`s never be awaited outside the context
+    /// of a step that comes afterwards -- doing so will cause an immediate
+    /// deadlock.
+    ///
+    /// Dropping a `StepHandle` has no effect on whether the step itself runs.
+    #[derive(Debug)]
+    pub struct StepHandle<T> {
+        #[pin]
+        receiver: oneshot::Receiver<T>,
+    }
+}
+
+// TODO: maybe export other receive methods from receiver, depending on need
+
+impl<T> Future for StepHandle<T> {
+    type Output = T;
+
+    fn poll(
+        self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Poll<Self::Output> {
+        let this = self.project();
+        this.receiver
+            .poll(cx)
+            .map(|output| output.expect("update-engine never drops the sender"))
```

Comment:
More precisely it's because `receiver` can never return None because we never drop the sender before sending a value from it. This reasoning is not local to the function, but is local to the update engine.

Reply:
This is perhaps ridiculously contrived, but I can trigger this panic with a very dumb modification to the example:

```diff
@@ -40,6 +40,10 @@ async fn main() {
         1_048_576,
     );

+    std::mem::drop(engine);
+    _ = download_handle_1.await;
+
+    /*
     // Write component 1 out to disk.
     context.register_write_step(&component_1, download_handle_1, false);

@@ -60,6 +64,7 @@ async fn main() {
     _ = display_handle.await;

     // Do not clean up the log file so people can inspect it.
+    */
 }
```

although if we do some kind of token / context with a lifetime tied to the engine, that would also make this impossible.

Reply:
Good catch -- yeah, the token addresses this issue. The only way to access a token is if:

* the previous step you're waiting on has completed.
* the engine has completed.

Both should tackle this issue. If we run into other cases in the future we should be able to address them separately, but I think (at least aspirationally) the panic is OK and presents the right API to the consumer.